File: lib/genocrunch_console/bin/get_version.py
#!/usr/bin/env python2.7
# -*- coding: utf-8 -*-
#get_version.py

import os
from subprocess import Popen, PIPE
from json import load, dump
from datetime import datetime

__dir__ = os.path.dirname(__file__)

pkg_list = []

# Get packages hard-coded in R scripts
sub = Popen([__dir__+'/get_version.sh'], stdout=PIPE, stderr=PIPE)
sub.wait()
pkg_list.extend([e for e in sub.stdout.read().split('\n') if e not in [None, '', ' ']])

# Get packages passed in etc json
json_fp = __dir__+'/../etc/genocrunchlib.json'
if os.path.exists(json_fp):
  data = load(open(json_fp))

for key, val in data['choices'].iteritems():
  pkg = [e['pkg'] for e in val if 'pkg' in e.keys() and e['pkg'] not in [None, '', ' ']]
  if len(pkg) > 0:
    pkg_list.extend(pkg)
pkg_list = list(set(pkg_list))

# Get version installed
sub = Popen([__dir__+'/get_version.R', ','.join(pkg_list)], stdout=PIPE, stderr=PIPE)
sub.wait()
versions = sub.stdout.read()

with open('version_'+str(datetime.now()).replace(' ', '_')+'.json', 'w') as f:
  f.write(versions)

exit()

File: lib/genocrunch_console/bin/get_version.R
#!/usr/bin/env Rscript
#get_version.R
# Get version of packages installed
args <- commandArgs(trailingOnly=FALSE)
pkgs <- unlist(strsplit(args[6], ','))

output <- c(1:(length(pkgs)+1))
output[1] <- paste('{"name":"R","version":"',
                   R.version.string,
                   '"}',
                   sep='')
for (i in 1:length(pkgs)) {
  if (!require(pkgs[i], character.only=TRUE)) {
    version <- 'NA'
  } else {
    version <- packageVersion(pkgs[i])
  }
  output[i+1] <- paste('{"name":"',
                       pkgs[i],
                       '","version":"',
                       version,
                       '"}',
                       sep='')
}
cat(paste('[', paste(output, collapse=','), ']', sep=''))

rm(list=ls())
